source/features/show-whitespaces.tsx

```diff
@@ -0,0 +1,99 @@
+import './show-whitespaces.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import onPrFileLoad from '../libs/on-pr-file-load';
+import onNewComments from '../libs/on-new-comments';
+
+const queue: Node[] = [];
+
+// Process a single line for each frame loop
+function loop(): void {
+	const line = queue.shift();
+
+	if (line === undefined) {
+		return;
+	}
+
+	const iterator = document.createNodeIterator(line, NodeFilter.SHOW_TEXT, {
+		acceptNode: node => {
+			if (node.childNodes.length === 0) {
+				return NodeFilter.FILTER_ACCEPT;
+			}
+
+			return NodeFilter.FILTER_REJECT;
+		}
+	});
+
+	const textNodes = [];
+	let node;
+
+	// Collect all nodes before modifying the root node anymore
+	while ((node = iterator.nextNode())) {
+		textNodes.push(node);
+	}
+
+	for (const textNode of textNodes) {
+		const nodeValue = textNode.textContent!;
+		if (nodeValue.length !== 0) {
+			const fragment = document.createDocumentFragment();
+
+			for (const char of nodeValue) {
+				if (char === '\t') {
+					fragment.append(<span className="pl-ws pl-tab">{char}</span>);
+				} else if (char === ' ') {
+					fragment.append(<span className="pl-ws pl-space">{char}</span>);
+				} else {
+					fragment.append(char);
+				}
+			}
+
+			(textNode as Element).replaceWith(fragment);
```

Comment:
What if instead of changing the dom, you just add a `data-rgh-spaces="••••"` attribute and then display them on all `.blob-code-inner` with `content: attr(data-rgh-spaces)`

Reply:
Done, please review at your convenience.